virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+        else:  # none virtio type, generate serial device directly
+            devices.append(qdevices.QDevice(serial_type, {"id": serial_id}))
+            devices[-1].set_param("reg", reg)
+        devices[-1].set_param('chardev', chardev_id)
+
+        return devices
+
+    def serials_define_by_params(self, serial_id, serial_params,
+                                 file_name=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param serial_params: serial params
+        :param file_name: the file path of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params['serial_type']
+        machine = serial_params.get('machine_type')
+
+        # Arm lists "isa-serial" as supported but can't use it,
+        # fallback to "-serial"
+        legacy_cmd = " -serial unix:'%s',server,nowait" % file_name
+        legacy_dev = qdevices.QStringDevice('SER-%s' % serial_id,
+                                            cmdline=legacy_cmd)
+        arm_serial = (serial_type == 'isa-serial'
+                      and 'arm' in serial_params.get("machine_type", ""))
+        if (arm_serial or not self.has_option("chardev")
+            or not self.has_device(serial_type)):
+            return legacy_dev
+
+        virtio_port_type = serial_params.get('virtio_port_type')
+        if virtio_port_type:
+            if "-mmio" in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
+
+        prefix = serial_params.get('virtio_port_name_prefix')
+        name = serial_id
+        name = prefix + name if prefix else name
+        serial_params['name'] = name
+
+        chardev_args = {'logfile': serial_params.get('chardev_logfile'),
+                        'logappend': serial_params.get('chardev_logappend')}
+        backend = serial_params.get('chardev_backend', 'unix_socket')
+        if backend not in ['ringbuf', 'spiceport', 'spicevmc']:
+            chardev_args.update({'mux': serial_params.get('chardev_mux')})
+        if backend in ['unix_socket', 'file', 'pipe', 'serial',
+                       'tty', 'parallel', 'parport']:
+            chardev_args.update({'path': file_name})
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_args.update({'debug': serial_params.get('chardev_debug'),
+                                 'name': name})
+        elif backend == 'stdio':
+            chardev_args.update({'signal': serial_params.get('chardev_signal')})
+        elif backend in ['udp', 'tcp_socket']:
+            chardev_args.update({'host': serial_params.get('chardev_host'),
```

Comment:
these  3 `elif` can be `if` ?

Reply:
@yanan-fu `elif` is more efficient than `if`